relays/lib-substrate-relay/src/messages_target.rs

```diff
@@ -102,27 +98,28 @@ where
 }
 
 #[async_trait]
-impl<SC, TC, P> TargetClient<P::MessageLane> for SubstrateMessagesTarget<SC, TC, P>
+impl<P> TargetClient<P::MessageLane> for SubstrateMessagesTarget<P>
 where
-	SC: Chain<
+	P: SubstrateMessageLane,
+	P::SourceChain: Chain<
 		Hash = <P::MessageLane as MessageLane>::SourceHeaderHash,
 		BlockNumber = <P::MessageLane as MessageLane>::SourceHeaderNumber,
 		Balance = <P::MessageLane as MessageLane>::SourceChainBalance,
 	>,
-	SC::Balance: TryFrom<TC::Balance> + Bounded,
-	TC: Chain<
+	<P::SourceChain as Chain>::Balance: TryFrom<<P::TargetChain as Chain>::Balance> + Bounded,
+	P::TargetChain: Chain<
 		Hash = <P::MessageLane as MessageLane>::TargetHeaderHash,
 		BlockNumber = <P::MessageLane as MessageLane>::TargetHeaderNumber,
 	>,
-	TC::Hash: Copy,
-	TC::BlockNumber: Copy,
-	TC::Header: DeserializeOwned,
-	TC::Index: DeserializeOwned,
-	<TC::Header as HeaderT>::Number: BlockNumberBase,
-	P: SubstrateMessageLane<SourceChain = SC, TargetChain = TC>,
+	<P::TargetChain as Chain>::Index: DeserializeOwned,
+	<P::TargetChain as BaseChain>::Hash: Copy,
+	<P::TargetChain as BaseChain>::BlockNumber: Copy,
+	<P::TargetChain as BaseChain>::Header: DeserializeOwned,
+	<<P::TargetChain as BaseChain>::Header as HeaderT>::Number: BlockNumberBase,
```

Comment:
To avoid repetition, maybe we could have type aliases that extract the right associated types? I.e.
```suggestion
	ChainIndexOf<P::TargetChain>: DeserializeOwned,
	BaseChainHashOf<P::TargetChain>: Copy,
	BaseChainBlockNumberOf<P::TargetChain>: Copy,
	BaseChainHeaderOf<P::TargetChain>: DeserializeOwned,
	BaseChainHeaderNumberOf<P::TargetChain>: BlockNumberBase,
```

Reply:
Not sure if we can do this in stable rust. Here is a minimal example https://play.rust-lang.org/ to illustrate the problem

Reply:
The ```
type ChainIndexOf<T> = <T as Chain>::Index;
```
should work fine, we don't need the extra bounds.

Reply:
It does work just fine. Still, I'm not a fan of replacing fully specified type names with type aliases. As someone just seeing the code for the first time it's just another layer of indirection. I have to learn to parse the type alias as rust. And I have to learn type names of the types in question since the naming can be ambiguous when one or more of the types have a multi word name. Also if the code base is to be uniform this adds a convention, for the sake of cca. 4 characters per replacement.